### ISSUE TITLE
Make colorized app icons optional

### ISSUE THREAD

**Issue by ninavizz** [QubesOS/qubes-issues#6652]
**The problem you're addressing (if any)**
Colorized app icons to communicate a qube's color, is a longstanding convention in Qubes OS. However, doing as much obfuscates the app's icon. Which is a non-trivial usability problem. The core "problem" the colorized app icons sought to solve for, is reinforcing with users which qube they are about to open an app into. Two different solutions to this problem [Bessie](https://www.qubes-os.org/attachment/survey/tour-bessie.webm) [Jackie](https://www.qubes-os.org/attachment/survey/tour-jackie.webm) developed for the [AppMenu Redesign](#5677) project.

**Describe the solution you'd like**
It had been my hope to eliminate colorized application icons altogether, in our updated AppMenu. However, enough users in the [AppMenu Redesign Survey](#6573) felt strongly against that. As such, a compromise to both meet the needs of longtime Qubes OS users desiring this feature, as well as newer users that find colorized app icons to be problematic, is to make the app icons optional. 

1. A global "Colorize app icons in menus" control in the Qubes Global Settings panel.
2. If users check this box to turn-on this feature, then the below setting would become enabled in each qube's individual Settings pane. I would suggest placing this control on the Applications tab, and not the first tab, so that as the user chooses different opacity values the icons colorize to that opacity in real-time... thus ensuring the legibility/coloration balance they desire.

![AppSettings-notips3](https://user-images.githubusercontent.com/8262612/118072031-dca40800-b35d-11eb-8a5b-d4a8e9c3ee86.png)

**Where is the value to a user, and who might that user be?**
For people who depend more upon "visual" information more than "written" information to make decisions from, it is frustrating to have the visual information of an icon obfuscated, as a solution for an unrelated problem that could be solved for, differently. 

As an example: when I go to open Signal, I look for the cyan ovular icon; when I go to open Firefox, I look for the fire-orange and electric-blue swirly-circle icon; GIMP, is black and white and an irregular shape. 

When the new AppMenu design is released, a different solution to the "Color my experience to the qube I'm opening a thing in!" problem can be solved, without depending upon creating another problem, to solve for that. 

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

Lots about how coloration is done of windows and icons, in Qubes OS, has been discussed in #2523. However, coloring over existing colored information, is not the only solution to the core problem colors seek to solve for, in the context of a user seeking to open an application from the appmenu.

**Comment by ninavizz** [comment]
In the recent AppMenu Redesign Survey, of 484 completed responses the question that received the strongest negative response to any of the questions, asked users if they liked that the app icons were not colorized per qube in each of the two prototypes presented. 

5.6% of users responded that they _Strongly Disliked_ the omission of colorized app icons.
13.9% responded that they simply _Disliked_ the omission of colorization. 

12.7% and 25.2% of users responded that they _Loved_ or _Liked_ the omission of app icon colorization, and 10.8% of respondents simply had not noticed the omission in the two prototype videos.

31.8% of all users felt _Neutral_ towards the omission. 

<img width="1321" alt="Screen Shot 2021-05-29 at 10 13 06 PM" src="https://user-images.githubusercontent.com/8262612/120093044-5c570400-c0cc-11eb-9e3d-8a3f30bb8071.png">

While the combined 19.5% _negative sentiment_ to removing the app icons was the strongest in the survey, the 31.8% Neutral was also the strongest _indifferent sentiment_ in the survey. A 37.9% _positive sentiment_, given the negative and the indifferent sentiment, has me hypothesizing that with a good alternative to reinforcing qube color in the new app menu, that in the use context of  selecting an app on one's own device, visual access to each app's uniquely colored icon will be desirable.  

A final note: Should we choose to launch with colorized selection bars in the new AppMenu (as suggested in the Bessie prototype), such a visually radical option would be made to be optional in a control panel for customizing options within the new AppMenu. So, yes, the feature this issue suggests, would effectively compliment that (and any additional colorized sub-menu pane decoration).

**Comment by GWeck** [comment]
Letting the user select icon coloriztion by VM is a good idea, because some VM colors and/or classes might be good for colorization, while other might be more expressive if the original colors are kept. For example, a red, untrustworthy VM might color its icons red in order to warn the user, while a template VM, usually black, might keep the original icon colors inorder to help the user find an application more easily.

On the other hand, it could be useful to provide a system-wide colorization option additionally, for users who would like to have everything look the same.

**Comment by marmarek** [comment]
Yes, global and per-qube options are totally possible. The interface would need to carry this information, though - we need three (four?) states in the qube settings:
 - enabled
 - disabled
 - using system default (which is: enabled/disabled)

Technical note:
In practice, I think the "global" option should really be set for a GUI domain, not really global. In most cases it is the same in practice (user have just one GUI domain - whether that is separate one, or dom0 doubling as such). But in the future, if we'll ever have multi-user option (with separate GUI domain per user), then it will become natural part of the user's settings.
Plus, it will be more convenient this way, as the icons are in fact handled in GUI domain itself. It doesn't change anything regarding the UI.

**Comment by ninavizz** [comment]
@GWeck I had not even thought of that, great idea! 

@marmarek I don't so much consider "Default" to be a state, however, with allthings involving policy files, I think surfacing to users whatever the "default" settings are, is important. Context matters a lot. Also, I have yet to get `4.1` on a test device to really poke-around to see what this GUI VM is like. Yet to procure a test device to enable that... but I look forward to seeing how everything looks in `4.1` just as soon as I can. :)

**Comment by marmarek** [comment]
> I don't so much consider "Default" to be a state

I mean, if we have some global knob, that can be overridden on per-qube basis, we have now those states of a per-qube setting:
1. it is controlled by the global knob (whatever its state is)
2. it is overridden for this qube, with state "enabled"
3. it is overridden for this qube, with state "disabled"

The final effect will be the same whether the thing is enabled via per-qube override, or via the global knob. But the distinction does matter when you toggle that global knob - for qubes with "controlled by the global knob" it will change their setting, but for qubes with setting specifically overridden, it will not.

That's the interaction of global setting and per-qube setting we use in various places already (see for example "networking" in qube settings - there is "default" value, which follows the value in global settings). It is totally possible to design it differently, if you have some specific interaction in mind.

**Comment by ninavizz** [comment]
@marmarek All helpful context—TY for sharing! I'm not sure the Network Settings stuff is 100% analagous, but I'll look closely at both, together, along with the rest of the per-qube vs global settings paradigms, while working on these.

**Comment by ffsec** [comment]
Suggestion: How about making this feature a bit more flexible by adding support for different icon decoration options?
Instead of having only one fixed option, maybe we could have a drop down list with decoration modes? This shouldn't be too much work to implement, as it would simply use a different script for each option behind the scenes.

So one option would be tinting the icons as it works right now.
Another option could be something like this (just and example):
![grafik](https://user-images.githubusercontent.com/31329006/120242367-3c524c80-c254-11eb-95ba-cacfafb1d0c7.png)
Of course any other decoration mode would be possible as well and could easily be implement as we go.

_Optional:
Organize decoration modes in their own sub folders, which contain the scripts, a small info file (containing option text and a description, possibly a version) and a preview image (an icon which has that mode applied), which is displayed next to the option.
That way you can automatically iterate through the sub folders and display all available decoration modes without having to adjust the window itself._

**Comment by ffsec** [comment]
Suggestion: How about making this feature a bit more flexible by adding support for different icon decoration options?
Instead of having only one fixed option, maybe we could have a drop down list with decoration modes? This shouldn't be too much work to implement, as it would simply use a different script for each option behind the scenes.

So one option would be tinting the icons as it works right now.
Another option could be something like this (just and example):
![grafik](https://user-images.githubusercontent.com/31329006/120242367-3c524c80-c254-11eb-95ba-cacfafb1d0c7.png)
Of course any other decoration mode would be possible as well and could easily be implement as we go.

_Optional:
Organize decoration modes in their own sub folders, which contain the scripts, a small info file (containing option text and a description, possibly a version) and a preview image (an icon which has that mode applied), which is displayed next to the option.
That way you can automatically iterate through the sub folders and display all available decoration modes without having to adjust the window itself._

Edit: At first I thought this would be a global option, which is applied to each vm that has the decoration option checked, but in theory it could be individually set for each vm. I guess, it all comes down to the available decoration modes. For example one could decorate some high risk vms with a more striking decoration style (which possibly trades icon quality for decoration visibility) and have a different decoration style for the other vms. Using the global decoration as default option would be good, while having the possibility to switch them for each individual vm.

**Comment by ninavizz** [comment]
@ffsec I like this idea, as a possible iteration to making any app icon decoration optional. However, it also misses the reason why the optional colorization is being pitched, in the first place. 

Important context: I began looking at what the colorized-icon feature exists to solve for in the first place, by considering the core design problem the existing feature sought to solve for:  "How to communicate the domain an app is about to be opened in or is running within through color." This core design problem I feel, should remain the focus, vs how to make the current solution in production suck less. If that makes sense. 

The new solution to the aforementioned design problem that is proposed with the App Menu redesign, is through colorizing other/related UI bits, while leaving the app icon entirely alone. The solutions you offer, are certainly an option—but will introduce other problems. Namely, that in a list-view it will create a very cluttered list for users to make their initial decision from. 

I'd like to get an updated app menu developed, first—and work to solve the core design problem, by looking at other areas of the UI not as loaded with semiotic details important to the user as app icons are. If that makes sense. If users remain unsatisfied with that, then I'm ok exploring an additional feature to iterate on the existing solution—but I don't want to remain locked-into one feature that is only one solution to the core problem. 

If you have thoughts on other ways to solve for the core design problem, I would love to hear them! I just really want to get away from the existing solution, before zooming-in on how to improve the existing solution—because the existing solution is so problematic, for the reasons expressed above. If that makes sense.

**Comment by ffsec** [comment]
@ninavizz
I don't think you can remove this feature completely. With color decorations in a custom app menu you could elimitate the icon decoration there, but for other areas it won't be possible at all.

I personally keep my frequently used applications on the desktop, placed in a certain order, so there is no easy way to eliminate icon decoration there. While in theory you could group the icons in colored boxes for each vm, it would completely remove the possiblity to place icons to the users preference. The same goes for 3rd party applications which display application icons.
So imo the icon decoration is an essential (security) feature in Qubes OS and it won't go anywhere. It just can't.

**Comment by ninavizz** [comment]
>  So imo the icon decoration is an essential (security) feature in Qubes OS and it won't go anywhere. It just can't.

Right. That's why this issue seeks to make it _optional_. For many users, it is a hinderance; whereas for others, it is embraced. 
The new app menu is likely to have a "Favorites" tab and possibly a "Recents" tab. In both, VMs and their colors are signified differently than colorized app icons (shown below). 

<img width="263" alt="Screen Shot 2021-06-01 at 9 29 08 AM" src="https://user-images.githubusercontent.com/8262612/120359335-69d2e080-c2bc-11eb-8d2d-c6129b1520f6.png">

As I'd said, the same "Problems" will always exist. They can be solved better, though, than taking information away from one place (the application's icon) to replace it with other information—and only in a fashion available to sighted users. I encourage you to follow the development of #5677 and to cite additional design problems you feel may have been overlooked and are in need of addressing, once alternate means to handle the primary two design problems are in place.

**Comment by ffsec** [comment]
@ninavizz
You keep referring to the custom app menu, which I'm not talking about. In that case you can get away with moving the icon decoration to other areas of the menu instead. However, there are more locations where an application can be launched from or where application icons are being used.
I personally use mainly the desktops to launch applications. The current icon tinting method has its limits but it's still better than nothing. Just and example from my dev machine:
![grafik](https://user-images.githubusercontent.com/31329006/120366248-f8685180-c2fe-11eb-8831-5fd4419e03bd.png)

With the other decoration method it looks like this:
![grafik](https://user-images.githubusercontent.com/31329006/120373285-b1328e80-c307-11eb-9adc-d5de48e7bcb1.png)

Windows for example is displaying an arrow on icons to signal that it's a shortcut, so I thought why not use a similar approach for the label color.
![grafik](https://user-images.githubusercontent.com/31329006/120372655-fb674000-c306-11eb-9379-9139f49a482b.png)

That being said, I'm not sure how you want to remove the label decoration from the icon itself in this case without messing with the desktop and its content itself. Now also keep in mind that there are many other desktop environments out there, which don't know anything about Qubes OS labels. The Qubes Team is not stupid, they know why they added the tinting to the icon in the first place.
(Taskbar, Task list, Tray icons, etc. not even mentioned yet)

So back to the main topic in here:
I think it would be way more useful to choose between different decoration modes rather than just having an on/off switch for tinting. Tinting doesn't work well and doesn't use accurate colors at all, so I can see why people might want to turn that off. But if you can have better decorations that actually display an accurate color without being too intrusive, I'd always opt for that.

**Comment by ffsec** [comment]
@ninavizz
I was not talking about the app menu itself, there are more locations where application icons are being used.
For example, what happens if you drag&drop an applications from the app menu onto the desktop, when decorations are disabled in the app menu? I think we need to make sure that desktop icons can/will have some sort of icon decoration, even if they are disabled in the app menu.

I guess there might be at least two sets of application icons needed. One with decoration and one without. Correct me if I'm wrong, but I think currently there is only one set of icons for each vm and they are automatically tinted? So if these become non-tinted, the coloration would be diabled globally and not just in the app menu.

Well yes, the icon tinting completely sucks, colors are not accurate at all and they lose a lot of details, so I can understand why that mode is not very popular.
![grafik](https://user-images.githubusercontent.com/31329006/120366248-f8685180-c2fe-11eb-8831-5fd4419e03bd.png)
I also don't like that mode and support getting rid of it, BUT I wouldn't want to get rid of icon decoration in general, as I consider it an essential security feature of Qubes OS. Luckily there are better ways that are quite usable and would likely be way more popular than the tinting.
![grafik](https://user-images.githubusercontent.com/31329006/120373285-b1328e80-c307-11eb-9adc-d5de48e7bcb1.png)

And last but not least, many users have their preferred desktop environment(s). So quite a large number of users will stick to those instead of using a new Qubes OS desktop enviroment, which (imo) seems to target mostly low-skilled users with few vms (office, browsing, nothing too fancy), rather than power users or IT professionals. So we need to make sure that the icons of one single app menu don't effect anything else.

**Comment by ninavizz** [comment]
@ffsec 
1. Again, this issue is to make colorization "optional." Not to outright eliminate it. 
2. In order to best support a goal of eventually eliminating it, I am keen to understand all the _design problems_ the icon colorization is sought to solve for with users, today. I identified two, up above in earlier comments—with the App Menu's functionality, being only one of them.
3. A design brief was written at the outset of the [App Menu](#5677) project (and is in its issue), clarifying its scope of reach. 

Should you read it, you'd see that, yes, this capability does seek to prioritize the needs of users new to Qubes. Which does not mean that it's ok to also make things suck for power-users; but, it does mean that we are seeking to optimize the adoption experience to reduce new user attrition, and to better support "regular folks" (not developers) with high security needs (so, Journalists, activists, people who are not Linux -ative users, etc).
 
I understand that you're not speaking to the App Menu. What would help me understand where your concerns, would be if you could simply form a "Design Problem" statement for each and every concern.  

**Example:** _"I want to see which VM an app belongs to, when I have many app icons on my desktop to open those apps from."_

Is the problem in the above cited use case the only other one you can think of, beyond the two I identified (App Menu and Task Bar buttons)—and supporting users in DEs other than the OEM configuration?

A _design problem statement_ does not factor in approaches to solutions, it only summarizes the problem. That keeps things open to understanding the breadth of solution opportunities to consider. The colorization "sucking" is not why I consider colorization to be a poor solution to all problems; the inadequacy of colorization as a solution, is that it removes some information to show other information—when both pools of visual information should be able to co-exist. I'm not trying to make colorization suck less; I'm trying to solve the problems it's currently embraced as a solution for, with _better solutions_.

Finally: None of this has been done, without proper research. In the survey of +880 users, most use an XFCE app menu or i3's DMenu. I will be publishing my formal research findings (from the survey and interviews with trainers), in the next week or two—but please do know they are forthcoming, and that none of this is being done in a vacuum. Below is what the users who participated in the AppMenu survey, said about how they use Qubes:

<img width="1010" alt="Screen Shot 2021-06-01 at 4 26 02 PM" src="https://user-images.githubusercontent.com/8262612/120402165-5cd2e300-c2f6-11eb-864b-461bd2aca06d.png">

**Comment by ffsec** [comment]
I give up. Good luck with whatever you are trying to accomplish, you will need it.

**Comment by alimirjamali** [comment]
I have been working on additional Icon Effects. The default tint, anti-aliased high quality Thin/Thick bordered icons, icons overplayed on VM label color, inverted icons for untrusted qube as well as untouched icons which I use for my highly trusted personal qube.

Project consist of a forked version of [qubesimgconverter](https://github.com/alimirjamali/personal-qubesos/blob/main/qubesimgconvertertt/README.md) python library with my additional effects, forked [qubesappmenus](https://github.com/alimirjamali/personal-qubesos/blob/main/qubesappmenustt/README.md) library and custom _qvm-appmenus-tt_ to apply  the per VM effect to the Appmenu. And a forked [icon-receiver](https://github.com/alimirjamali/personal-qubesos/blob/main/window-icon-updater-tt/README.md) daemon to apply per VM effect to the icons of running applications.

The tools look for an special key in VM features (i.e. `ttfilter`) to apply custom icon effect per VM. I have been using it during the last few couple of weeks personally and I am happy with how it turned out. Screenshots are available on the Github README.md pages I shared above. Here is a sneak peek (two tinted Firefox icons, two thin-bordered ones, one overlay, one untouched (Firefox from my personal qube) and one inverted paranoid icon (Firefox of untrusted qube):

![image](https://github.com/QubesOS/qubes-issues/assets/5407317/034d4d5a-cedd-4d1f-a5c9-28b9ac4e1d9c)

Additional idea would be alpha-compositing small VM Icon on bottom right of the Application icons. I have written the alpha compositor for the qubesimgconverter library (used for high quality thin-borders). But I am looking for more ideas.

I am working on other aspects of the project. Like adding the ability to send App Window of qube to the desired Workspace (e.g. by have a `ttworkspace` tag in target VM features and getting PID from qvm_run subprocess. I am also working on rewriting my [custom label maker](https://github.com/alimirjamali/personal-qubesos/blob/main/qubes-label-tt/README.md) in Python. Once the project is fully finished, Hopefully I could submit it as a potential community project in a separate Github issue.
